tests/purs/passing/DerivingFunctor.purs

```diff
@@ -69,6 +69,35 @@ maTests = do
   assert' "map - M5" $ map show m5L == m5R
   assert' "map - M6" $ map show m6L == m6R
 
+data Fun1 a = Fun1 (Int -> Int -> a)
+derive instance Functor Fun1
+
+f1Test = do
+  assert' "map - Fun1" do
+    let
+      fn = show
+      left a b = a + b
+      right a b = fn $ left a b
+      Fun1 left' = map fn $ Fun1 left
+    left' 1 2 == right 1 2
+
+data Fun2 a = Fun2 ((a -> a -> Int) -> Int)
+derive instance Functor Fun2
```

Comment:
Ah, right, the double-contravariant special handling of functions.

I'm not hard against having this, but I will note the following:
* `Functor` deriving has gotten along without it so far
* It has nothing to do with `Foldable` and `Traversable`
* Once we add it, removing or possibly changing it would be a breaking change

Why would we remove or change it? Really, there's no reason `Function` should be special in this regard; one `Profunctor` is as valid as another. The general case might entail expanding `Functor`-deriving to any arguments of types that implement `Functor`, `Bifunctor`, `Contravariant`, or `Profunctor`, with appropriate handling of variance and appropriate positioning of the argument.

But I figure that's probably a distraction from this PR, so I'm wondering if it's better to keep this PR simple(r) and leave `Functor` deriving alone for now.

Reply:
Ah, good points. That will definitely make the tests easier to write.